### PR TITLE
impl(pubsub): expose subscription and ack id on the ack handler

### PR DESCRIPTION
**Pull request by alevenberg**
Part of [#13725](https://github.com/googleapis/google-cloud-cpp/issues/13725)

This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13792)

### CODE REVIEW

File: google/cloud/pubsub/exactly_once_ack_handler.h [hunk 1]
```diff
@@ -122,6 +123,16 @@ class ExactlyOnceAckHandler {
     }
     /// The implementation for `ExactlyOnceAckHandler::delivery_attempt()`
     virtual std::int32_t delivery_attempt() const { return 0; }
+    /// Returns the ack id for the handler. There is no corresponding public
+    /// interface to access the ack id in `AckHandler`. This is for internal
```
Comment (devbww) [inline google/cloud/pubsub/exactly_once_ack_handler.h:127]: s/AckHandler/ExactlyOnceAckHandler/?

File: google/cloud/pubsub/exactly_once_ack_handler.h [hunk 2]
```diff
@@ -122,6 +123,16 @@ class ExactlyOnceAckHandler {
...
+    /// use only.
+    virtual std::string ack_id() const { return ""; }
+    /// Returns the subscription for the handler. There is no corresponding
+    /// public interface to access the subscription id in `AckHandler`. This
```
Comment (devbww) [inline google/cloud/pubsub/exactly_once_ack_handler.h:131]: Ditto?

**Review by devbww — APPROVED** [review]

File: google/cloud/pubsub/exactly_once_ack_handler.h (hunk 1 above)
Comment (alevenberg) [reply google/cloud/pubsub/exactly_once_ack_handler.h:127]: Good catch. Done

File: google/cloud/pubsub/exactly_once_ack_handler.h (hunk 2 above)
Comment (alevenberg) [reply google/cloud/pubsub/exactly_once_ack_handler.h:131]: Done

File: google/cloud/pubsub/mocks/mock_exactly_once_ack_handler.h [hunk 3]
```diff
@@ -37,8 +37,9 @@ class MockExactlyOnceAckHandler : public pubsub::ExactlyOnceAckHandler::Impl {
  public:
   MOCK_METHOD(future<Status>, ack, (), (override));
   MOCK_METHOD(future<Status>, nack, (), (override));
-  MOCK_METHOD(std::string, ack_id, ());
```
[the remark this answers is not in the document]
Comment (alevenberg) [reply google/cloud/pubsub/mocks/mock_exactly_once_ack_handler.h:40]: These are good points. Let me revist the PR and make some changes.

File: google/cloud/pubsub/mocks/mock_exactly_once_ack_handler.h [hunk 4]
```diff
@@ -37,8 +37,9 @@ class MockExactlyOnceAckHandler : public pubsub::ExactlyOnceAckHandler::Impl {
...
   MOCK_METHOD(std::int32_t, delivery_attempt, (), (const, override));
+  MOCK_METHOD(std::string, ack_id, (), (const, override));
+  MOCK_METHOD(pubsub::Subscription, subscription, (), (const, override));
```
[the remark this answers is not in the document]
Comment (alevenberg) [reply google/cloud/pubsub/mocks/mock_exactly_once_ack_handler.h:42]: Mocks for private methods need to be public https://google.github.io/googletest/gmock_cook_book.html

I will leave these mocks alone and make impl only mocks.

[the remark this answers is not in the document]
Comment (alevenberg) [reply google/cloud/pubsub/mocks/mock_exactly_once_ack_handler.h]: Now it is not changing anything public. I might need to revisit this later, but this works for now.

**Pull request merged**